### PR TITLE
HOCS-2563 : add countries and territories files as local resources.

### PR DESCRIPTION
**Pull request by abarker66**
This solution for the removed countries and territories files from the gov.uk registers uses direct copies installed as local resources and served via the rest controller as though it were an external resource, this allows us to us the same object mapping services provided by Spring/Jackson. The application.properties now references the files as local resources.

### CODE REVIEW

File: src/test/resources/application-test.properties [hunk 1]
```diff
@@ -22,8 +22,8 @@ api.scottish.parliament=https://data.parliament.scot/api/members
 api.ni.assembly=http://data.niassembly.gov.uk/members.asmx/GetAllCurrentMembers
 api.european.parliament=http://www.europarl.europa.eu/meps/en/download/advanced/xml?countryCode=GB
 api.welsh.assembly=https://senedd.assembly.wales/mgwebservice.asmx/GetCouncillorsByWard
-api.country.register=https://country.register.gov.uk/records.json?page-size=5000
-api.territory.register=https://territory.register.gov.uk/records.json?page-size=5000
+api.country.register=http://localhost:8085/countries/local
```
Comment (zuzak) [inline src/test/resources/application-test.properties:25]: We can't hardcode the port number here, as it's variable.

File: src/main/resources/territories-list.json [hunk 2]
```diff
@@ -0,0 +1 @@
..."14","entry-timestamp":"2016-12-15T12:15:07Z","key":"BAT","item":[{"official-name":"British Antarctic Territory","name":"British Antarctic Territory","territory":"BAT"}]},"PF":{"index-entry-number":"28","entry-number":"28","entry-timestamp":"2016-12-15T12:15:07Z","key":"PF","item":[{"official-name":"French Polynesia","name":"French Polynesia","territory":"PF"}]},"TK":{"index-entry-number":"72","entry-number":"72","entry-timestamp":"2016-12-15T12:15:07Z","key":"TK","item":[{"official-name":"Tokelau","name":"Tokelau","territory":"TK"}]},"PM":{"index-entry-number":"65","entry-number":"65","entry-timestamp":"2016-12-15T12:15:07Z","key":"PM","item":[{"official-name":"Saint Pierre and Miquelon","name":"Saint Pierre and Miquelon","territory":"PM"}]},"PN":{"index-entry-number":"57","entry-number":"57","entry-timestamp":"2016-12-15T12:15:07Z","key":"PN","item":[{"official-name":"Pitcairn, Henderson, Ducie and Oeno Islands","name":"Pitcairn, Henderson, Ducie and Oeno Islands","territory":"PN"}]},"UM-71":{"index-entry-number":"48","entry-number":"48","entry-timestamp":"2016-12-15T12:15:07Z","key":"UM-71","item":[{"official-name":"Midway Islands","name":"Midway Islands","territory":"UM-71"}]}}
```
Comment (zuzak) [inline src/main/resources/territories-list.json:1]: Probably worth running these through a JSON pretty-printer. Everything being on one line will make it very difficult to review changes in future.

File: src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java [hunk 3]
```diff
@@ -27,6 +33,34 @@ public CountryResource(CountryService countryService) {
...
+
+    @GetMapping(value = "/territories/local", produces = APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<?> getAllActiveTerritories() {
+
+        log.info("getAllActiveTerritories - local resource");
+        File territories = null;
+        try {
+            territories = new ClassPathResource("territories-list.json").getFile();
+        } catch (IOException e) {
+            log.info("local territories file not found.");
```
Comment (zuzak) [inline src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java:58]: should this be higher than `info`?

File: src/main/resources/territories-list.json (hunk 2 above)
Comment (abarker66) [reply src/main/resources/territories-list.json:1]: we would not review these file but just replace them, they are big and not for humans.

File: src/test/resources/application-test.properties (hunk 1 above)
Comment (abarker66) [reply src/test/resources/application-test.properties:25]: How do I fix that?

**Comment by zuzak** [comment]
We need to check we comply with the terms of the OGL before merging this: we can't just copy/paste code. Might need to attribute inside the `LICENSE` file?

File: src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java (hunk 3 above)
Comment (abarker66) [reply src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java:58]: it is DECS convention, what do you think would be appropriate.

File: src/test/resources/application-test.properties (hunk 1 above)
Comment (abarker66) [reply src/test/resources/application-test.properties:25]: there are a few references to fixed port localhost in this and other properties files - how is this different?

Comment (abarker66) [reply src/test/resources/application-test.properties:25]: I do not know how to fix this. The URL (URI really) is referencing its own Rest Controller, so how do I then reference localhost with the correct port?? do we need some clever config to do this? or should I just reference the file in github like I did in the first solution?

Comment (abarker66) [reply src/test/resources/application-test.properties:25]: I misread this : I have fixed the url with using the server.port variable.

File: src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java (hunk 3 above)
Comment (abarker66) [reply src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java:58]: The failures are now logged at error level. I should point out that there was no logging at all in this class previously.

File: README.md [hunk 4]
```diff
@@ -122,3 +122,5 @@ This project is authored by the Home Office.
 ### License 
 
 This project is licensed under the MIT license. For details please see [License](LICENSE) 
+
+This project Contains public sector information licensed under the Open Government Licence v3.0. (http://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/)
```
Comment (zuzak) [inline README.md:126]:
```suggestion
This project contains public sector information licensed under the Open Government Licence v3.0. (http://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/)
```

Add this to the LICENSE file too. Just a "Contains public sector information licensed under the Open Government Licence v3.0".

File: src/test/resources/application-test.properties (hunk 1 above)
Comment (LiamMacP) [reply src/test/resources/application-test.properties]: Change to use the port variable have been completed.

File: src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java (hunk 3 above)
Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/info/api/CountryResource.java]: We do need a review of the the logging levels. I'm happy with this been logged as an error - to which it is.

File: src/main/resources/territories-list.json (hunk 2 above)
Comment (LiamMacP) [reply src/main/resources/territories-list.json]: JSON files have been ran through a prettier - change implemented.

File: LICENSE [hunk 5]
```diff
@@ -0,0 +1,24 @@
+MIT License
+
+Copyright (c) 2018
```
Comment (LiamMacP) [inline LICENSE:3]: We should review this - as we use Crown Copyright which is special - I will add a ticket to the backlog to review this.

**Review by LiamMacP — APPROVED** [review]
Changed look good to me. License file have been added and OGL 3.0 reference for the data brought in. Appropriate changes have been made to test properties to allow this to work.

**Pull request merged**